amqp/src/MessageBusAmqp.cpp

```diff
@@ -16,68 +16,183 @@
+    if (msgSent != DeliveryState::Accepted) {
+        logError("Message sent (Rejected)");
+        return fty::unexpected(msgSent);
     }
 
-    // Send request
-    return m_busAmqp->request(msg, timeOut);
+    logDebug("Message sent (Accepted)");
+    return {};
+}
+
+fty::Expected<Message, DeliveryState> MessageBusAmqp::request(const Message& message, int timeoutInSeconds) noexcept
+{
+    try {
+        if (!isServiceAvailable()) {
+            logDebug("Service not available");
+            return fty::unexpected(DeliveryState::Unavailable);
+        }
+
+        // Sanity check
+        if (!message.isValidMessage()) {
+            return fty::unexpected(DeliveryState::Rejected);
+        }
+        if (!message.needReply()) {
+            return fty::unexpected(DeliveryState::Rejected);
+        }
+
+        logDebug("Synchronous request and checking answer until {} second(s)...", timeoutInSeconds);
+        proton::message msgToSend = getAmqpMessage(message);
+
+        // Promise and future to check if the answer arrive constraint by timeout.
+        Promise<Message> promiseSyncRequest(*this, msgToSend.reply_to());
+
+        bool msgArrived = false;
+
+        auto msgReceived = receive(
+            msgToSend.reply_to(),
+            std::move(std::bind(&Promise<Message>::setValue, &promiseSyncRequest, std::placeholders::_1)),
+            proton::to_string(msgToSend.correlation_id()));
+        if (!msgReceived) {
+            return fty::unexpected(DeliveryState::Aborted);
+        }
+
+        // Send message
+        auto msgSent = send(message);
+        if (!msgSent) {
+            return fty::unexpected(DeliveryState::Aborted);
+        }
+
+        // Wait response
+        if (promiseSyncRequest.waitFor(timeoutInSeconds * 1000)) {
+            msgArrived = true;
+        }
+
+        // Unreceive filter
+        auto unreceivedFilter = m_clientPtr->unreceiveFilter(proton::to_string(msgToSend.correlation_id()));
+        if (unreceivedFilter != DeliveryState::Accepted) {
+            logWarn("Issue on unreceive filter");
+        }
+
+        if (!msgArrived) {
+            logError("No message arrived within {} seconds!", timeoutInSeconds);
+            return fty::unexpected(DeliveryState::Timeout);
+        }
+
+        auto value = promiseSyncRequest.getValue();
+        if (value) {
+            return *value;
+        }
+        return fty::unexpected(DeliveryState::Aborted);
+    } catch (const std::exception& e) {
+        logError("Exception in synchronous request: {}", e.what());
+        return fty::unexpected(DeliveryState::Aborted);
+    }
 }
 
 void MessageBusAmqp::setConnectionErrorListener(ConnectionErrorListener errorListener)
 {
-    m_busAmqp->setConnectionErrorListener(errorListener);
+    m_clientPtr->setConnectionErrorListener(errorListener);
 }
 
 const std::string& MessageBusAmqp::clientName() const noexcept
 {
-    return m_busAmqp->clientName();
+    return m_clientName;
```

Comment:
I think that it complicate a little the code but the first reason that I wanted to have the same class for Promise in common. With this class, it would be impossible ...

Reply:
ok, but you remove this functionnality